src/System.Threading.Tasks.Extensions/src/System/Threading/Tasks/ValueTask.cs

```diff
@@ -0,0 +1,177 @@
+    /// data instead of one, which is more data to copy.  It also means that if a method that returns one of these
+    /// is awaited within an async method, the state machine for that async method will be larger due to needing
+    /// to store the struct that's two fields instead of a single reference.
+    /// </para>
+    /// <para>
+    /// Further, for uses other than consuming the result of an asynchronous operation via await, 
+    /// <see cref="ValueTask{TResult}"/> can lead to a more convoluted programming model, which can in turn actually 
+    /// lead to more allocations.  For example, consider a method that could return either a <see cref="Task{TResult}"/> 
+    /// with a cached task as a common result or a <see cref="ValueTask{TResult}"/>.  If the consumer of the result 
+    /// wants to use it as a <see cref="Task{TResult}"/>, such as to use with in methods like Task.WhenAll and Task.WhenAny, 
+    /// the <see cref="ValueTask{TResult}"/> would first need to be converted into a <see cref="Task{TResult}"/> using 
+    /// <see cref="ValueTask{TResult}.AsTask"/>, which leads to an allocation that would have been avoided if a cached 
+    /// <see cref="Task{TResult}"/> had been used in the first place.
+    /// </para>
+    /// <para>
+    /// As such, the default choice for any asynchronous method should be to return a <see cref="Task"/> or 
+    /// <see cref="Task{TResult}"/>. Only if performance analysis proves it worthwhile should a <see cref="ValueTask{TResult}"/> 
+    /// be used instead of <see cref="Task{TResult}"/>.  There is no non-generic version of <see cref="ValueTask{TResult}"/> 
+    /// as the Task.CompletedTask property may be used to hand back a successfully completed singleton in the case where
+    /// a <see cref="Task"/>-returning method completes synchronously and successfully.
+    /// </para>
+    /// </remarks>
+    [StructLayout(LayoutKind.Auto)]
+    public struct ValueTask<TResult> : IEquatable<ValueTask<TResult>>
+    {
+        /// <summary>The task to be used if the operation completed asynchronously or if it completed synchronously but non-successfully.</summary>
+        internal readonly Task<TResult> _task;
+        /// <summary>The result to be used if the operation completed successfully synchronously.</summary>
+        internal readonly TResult _result;
+
+        /// <summary>Initialize the <see cref="ValueTask{TResult}"/> with the result of the successful operation.</summary>
+        /// <param name="result">The result.</param>
+        public ValueTask(TResult result)
+        {
+            _task = null;
+            _result = result;
+        }
+
+        /// <summary>
+        /// Initialize the <see cref="ValueTask{TResult}"/> with a <see cref="Task{TResult}"/> that represents the operation.
+        /// </summary>
+        /// <param name="task">The task.</param>
+        public ValueTask(Task<TResult> task)
+        {
+            _task = task;
```

Comment:
The Task will already be null if one uses ```default(ValueTask<TResult>)```, which would be the same as ```new ValueTask(default(TResult))```.  It's not clear to me why ```new ValueTask((Task<TResult>)null)``` should be different in terms of throwing.  There's not a lot of cost to a branch, but it's not free, and if the goal of this type is to help minimize as much cost as possible, it seems odd that we'd add such overheads in to prevent a subset of "incorrect" usage.  You could even make the argument that specifying a null task is the same as saying "I don't want to wait for anything".

I don't feel strongly about it, and I'm mostly playing devil's advocate.  I just don't know what real value adding such a check would provide.  What real and dangerous situation is it helping to prevent?